Better UI/UX for suggestions for new users (IOS-157)

zeitschlag opened the pull request:
Say hello to the new "Hey, do you already follow there people?"-screen, that appears every time you open the app and don't follow people. And when you press the "Find people to follow"-button on the empty-screen

It uses the new `UserView` for every account, but offers the option to follow them all.

<img src="https://github.com/mastodon/mastodon-ios/assets/2580019/734ccfba-9133-41be-a84b-c0b46e676213" width="300px"></img><img src="https://github.com/mastodon/mastodon-ios/assets/2580019/6f2b756f-c9b7-4aad-82d3-33a829c3d142" width="300px"></img>

zeitschlag committed:
Minor UI-improvements in UserView (IOS-157)

Make avatar-image bigger, fix position of follow-button

zeitschlag committed:
Remove collection view from suggestion-screen (IOS-157)

zeitschlag committed:
Cleanup/Style TableView (IOS-157)

zeitschlag committed:
Show only five accounts (IOS-157)

DK1DA commented on Mastodon/Scene/SuggestionAccount/TableViewCell/SuggestionAccountTableViewCell+ViewModel.swift:
```diff
@@ -138,4 +74,62 @@ extension SuggestionAccountTableViewCell {
     }
+
+    class ViewModel {
+        var disposeBag = Set<AnyCancellable>()
```
Can this be made `private`?
```suggestion
        private var disposeBag = Set<AnyCancellable>()
```

DK1DA commented on Mastodon/Scene/SuggestionAccount/TableViewCell/SuggestionAccountTableViewCell+ViewModel.swift:
```diff
@@ -138,4 +74,62 @@ extension SuggestionAccountTableViewCell {
+        
+        @Published public var userIdentifier: UserIdentifier?       // me
+        
+        @Published var avatarImageURL: URL?
```
Why is this `internal` whereas `authorName` and `authorUsername` are private? Shouldn't it be on the same access level?

zeitschlag replied:
To be honest: I don't know (yet), I didn't write this code, I just moved it around a bit :)

kimar reviewed (commented):
Great work. 👏  I see that your Pr is a draft, just adding a few notes here.

zeitschlag committed:
Add request-follow/pending-states to follow-button (IOS-157, IOS-140)

DK1DA commented on Mastodon/Scene/Share/View/TableviewCell/UserTableViewCell+ViewModel.swift:
```diff
@@ -52,20 +54,24 @@ extension UserTableViewCell {
                     self?.userView.setButtonState(.unfollow)
+                } else if requested.contains(user.id) {
+                    self?.userView.setButtonState(.pending)
+                } else if user.locked {
```
🥇 thank you @zeitschlag !

DK1DA commented on Mastodon/Diffable/Search/SearchHistorySection.swift:
```diff
@@ -32,7 +32,12 @@ extension SearchHistorySection {
+                        value: user,
+                        followedUsers: authContext.mastodonAuthenticationBox.inMemoryCache.$followingUserIds.eraseToAnyPublisher(),
+                        blockedUsers: authContext.mastodonAuthenticationBox.inMemoryCache.$blockedUserIds.eraseToAnyPublisher(),
+                        followRequestedUsers: authContext.mastodonAuthenticationBox.inMemoryCache.$followRequestedUserIDs.eraseToAnyPublisher()
```
We'd also need to fetch the pending requests from https://docs.joinmastodon.org/methods/follow_requests/ to pre-seed this e.g. on user switch etc. basically the same as done here: https://github.com/mastodon/mastodon-ios/blob/develop/MastodonSDK/Sources/MastodonCore/Service/AuthenticationService.swift#L50

zeitschlag replied:
Ah, you're right! Good that you saw this 👍  I'll add that 🙂

zeitschlag committed:
Use UserView etc. to present suggested people to follow (IOS-157)

zeitschlag committed:
Extract follow-button into its own file (IOS-157)

zeitschlag committed:
Add footer-view to follow all suggested accounts (IOS-157)

- Also: Rename folder as this contains not only the cell anymore
- This is just UI, follow-function is still missing

zeitschlag committed:
Minor UI-fixes (IOS-157)

- NavigationBar
- ContentInset for TableView

zeitschlag committed:
Update title for suggestion-screen (IOS-157)

zeitschlag committed:
Follow all suggested user accounts (IOS-157)

zeitschlag committed:
Make follow work (IOS-157)

The secret is to set the delegate before getting the account. This doesn't feel right.

zeitschlag committed:
Re-introduce viewModel for each cell and set button-state (IOS-157)

We need this viewModel for the button--state of the follow-button.
There's still a runtime-issue that resets the button-state to follow.

zeitschlag committed:
Download (and cache) requested follows (IOS-157)

Thanks to @kimar for pointing this out!

zeitschlag committed:
Kill some log-messages along the way (IOS-157)

zeitschlag committed:
Add emoji-support (IOS-157)

zeitschlag committed:
Sprinkle in some localization (IOS-157)

zeitschlag committed:
Make button-state work (IOS-157)

Reason for button-state not working/updating feels like to be a weird combination of Combine, UIKit, Snapshots, CoreData and me being stupid and not getting a hang on it.

zeitschlag committed:
Present suggestions when user visits onboarding and doesn't follow people (IOS-157)

zeitschlag committed:
Don't set title for profile-screen when modal (IOS-157)

aka when account-suggestion-screen

zeitschlag committed:
Fix navigation-bar-style (IOS-157)

zeitschlag committed:
Fix build (IOS-157)

zeitschlag committed:
Dismiss after following all (IOS-157)

zeitschlag committed:
Refresh hometimeline (IOS-157)

zeitschlag committed:
Fix Suggestion-screen on iPad (IOS-157)

Follow-button wasn't attached to the right side

zeitschlag committed:
Fix name for development (IOS-157)

zeitschlag committed:
Fix search-button on empty-view (IOS-157)

Wasn't tappable on iPhone

DK1DA commented on MastodonSDK/Sources/MastodonUI/View/Content/UserView.swift:
```diff
@@ -99,7 +99,13 @@ public final class UserView: UIView {
     }()
-    
+
+    private let followButtonWrapper = {
```
💡  Looks like we could inline this after all...
```suggestion
    private let followButtonWrapper = UIView()
```

zeitschlag replied:
Oh 🙈😅 You're absolutely right!

DK1DA commented on Mastodon/Scene/SuggestionAccount/SuggestionAccountViewController.swift:
```diff
@@ -168,46 +98,37 @@ extension SuggestionAccountViewController: UITableViewDelegate {
+extension SuggestionAccountViewController: SuggestionAccountTableViewFooterDelegate {
+    func followAll(_ footerView: SuggestionAccountTableViewFooter) {
+        viewModel.followAllSuggestedAccounts(self) {
+            OperationQueue.main.addOperation {
```
💡  No need to change, but I think we're mainly using GCD directly 👇 
```suggestion
            DispatchQueue.main.async {
```

zeitschlag replied:
Great, that you saw that. I'm the only one using `OperationQueue.main.addOperation`. I changed there in the other places, too 👍

kimar reviewed (approved):
LGTM 👍

zeitschlag committed:
Simplify overly complicated init (IOS-157)

Thank you @kimar!

zeitschlag committed:
Use same way to run stuff on main everywhere (IOS-157)
